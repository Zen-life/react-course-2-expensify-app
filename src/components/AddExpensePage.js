import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense} from '../actions/expenses';

//the 'export' class enables testing this unconnect class, 
// rather than the export default one
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {                
        this.props.addExpense(expense); 
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}


const mapDispatchToProps = (dispatch) => ({
    // the goal is to return an object
    // inside return, we can define various props to call dispatch
    addExpense: (expense) => dispatch(addExpense(expense))
});
    
export default connect(undefined, mapDispatchToProps)(AddExpensePage);