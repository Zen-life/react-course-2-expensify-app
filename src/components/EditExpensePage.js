import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
 
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/dashboard');
    };

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/dashboard');
    };
    render() {
        return (
            <div> 
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>                
                </div>
                <div className="content-container">
                    <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                    />

                    <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
                </div>              
            </div>
        );
    }
};



// we need to search for expenses whose id match 'props.match.params.id'
// to access the details we need access to props, hence the 2nd props arg below
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)        
    }
};


    // the goal is to return an object
    // inside return, we can define various props to call dispatch
    // the props is required (according to Insctructor) to pass data returned by the Object in the startRemoveExpense() above
    // Although my test showed it works fine without the passing id and expense.
const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});



export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);