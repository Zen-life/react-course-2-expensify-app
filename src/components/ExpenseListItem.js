import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';



// same as below but destructured here
// parse createdAt to moment and set format. amount to numeral
const ExpenseListItem = ({description, amount, createdAt, id}) => (
    <div>
        <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
        <p>
        {numeral(amount / 100).format('0,0.00')} 
        - 
        {moment(createdAt).format('MMMM Do, YYYY')} 
        </p>        
    </div>
);

/*
const ExpenseListItem = (props) => (
    <div>
        <h3>{props.description}</h3>
        <p>{props.amount} - {props.createdAt}</p>

    </div>
);
*/



export default ExpenseListItem;