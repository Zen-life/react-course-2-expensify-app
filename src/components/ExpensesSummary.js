import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import SelectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';


// destructured these variables: expenseCount, expensesTotal
export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount <= 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('0,0.00');

    return(
        <div>
        <p>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}</p>
        </div>
    );
};



const mapStatetoProps = (state) => {
    const visibleExpenses = SelectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

// the below is norm way of mapping state to props
export default connect(mapStatetoProps)(ExpensesSummary);
