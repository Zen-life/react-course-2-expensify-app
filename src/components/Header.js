import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
    // this is a regular html5 header tag
    //  exact={true} is del as no longer linking to the root '/'
    <header className="header">
        <div className="content-container">  
            <div className="header__content">
                <Link className="header__title" to="/dashboard">
                    <h1>Expensify</h1>
                </Link>
                <button className="button button--link" onClick={startLogout}>Logout</button>
            </div>
        </div> 
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

// no state required on this component so undefined for first arg.
export default connect(undefined, mapDispatchToProps)(Header);