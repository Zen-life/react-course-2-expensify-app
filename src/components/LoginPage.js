import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__box">
        <h1 className="box-layout__title">Expensify</h1>
        <p>it's time to get your expenses under control</p>
        <button className="button" onClick={startLogin}>Login with Google</button>
        </div>
        
    </div>
);
    
// the startLogin prop in the object, will dispatch the startlogin action()
const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

// exporting the connected version of the component
// 'undefined' represent mapStateToProps, which we don't need this time
// (LoginPage) is calling the mapDispatchToProps connect and parse in the
// component 'LoginPage' that we want to connect
export default connect(undefined, mapDispatchToProps)(LoginPage);