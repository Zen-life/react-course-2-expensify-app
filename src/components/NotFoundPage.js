import React from 'react';
import { Link } from 'react-router-dom';

const NotFoundPage = () => (
    <div>
        404! This page is not found.
        <Link to="/dashboard">Go home</Link>
    </div>
);

export default NotFoundPage;