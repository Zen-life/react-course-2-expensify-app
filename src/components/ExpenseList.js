import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import SelectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? ( 
                <p>No Expenses</p> 
                ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />; 
                })
            )
        }
    </div>
);
const mapStatetoProps = (state) => {
    return {
        expenses: SelectExpenses(state.expenses, state.filters)
    };
};

// the below is norm way of mapping state to props
export default connect(mapStatetoProps)(ExpenseList);
