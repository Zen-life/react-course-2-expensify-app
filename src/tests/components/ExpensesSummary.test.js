import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';


test('Should correctly render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={195} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should correctly render ExpensesSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={13} expensesTotal={21348765} />);
    expect(wrapper).toMatchSnapshot();
});