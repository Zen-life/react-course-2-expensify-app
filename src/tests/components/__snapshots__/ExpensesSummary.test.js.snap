// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    1.95
  </p>
</div>
`;

exports[`Should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    13
     
    expenses
     totalling 
    213,487.65
  </p>
</div>
`;
